pkg/io/providers/aws_test.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"context"
+	"testing"
+
+	awshttp "github.com/aws/aws-sdk-go-v2/aws/transport/http"
+)
+
+func Test_newS3Client(t *testing.T) {
+	tests := []struct {
+		name  string
+		creds s3Credentials
+	}{
+		{
+			name: "only accesskey and secretkey set",
+			creds: s3Credentials{
+				AccessKey: "foo",
+				SecretKey: "bar",
+			},
+		},
+		{
+			name: "all options set ",
+			creds: s3Credentials{
+				AccessKey:        "foo",
+				SecretKey:        "bar",
+				Endpoint:         "https://foobar.com",
+				Region:           "eu01",
+				Insecure:         true,
+				S3ForcePathStyle: true,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			got, err := newS3Client(ctx, &tt.creds)
+			if err != nil {
+				t.Errorf("newS3Client() error = %v,", err)
+				return
+			}
+			s3opts := got.Options()
+			awsCreds, err := s3opts.Credentials.Retrieve(ctx)
+			if err != nil {
+				t.Errorf("get aws credentials error = %v,", err)
+				return
+			}
+			if awsCreds.AccessKeyID != tt.creds.AccessKey {
```

Comment:
I guess we could just check that the `CredentialsProvider` is of type `StaticCredentialsProvider` and remove the other assertion of AccessKey and SecretKey
```go
if tt.creds.AccessKey != "" && tt.creds.SecretKey != "" {
	_, ok := s3opts.Credentials.(credentials.StaticCredentialsProvider)
	if !ok {
		t.Errorf("credentials should be StaticCredentialsProvider, got %T,", s3opts.Credentials)
		return
	}
}
``` 
That would prevent accidental leakage of credentials

Reply:
It's a test, what credentials could possibly be leaked here? My comment is more that it's a lot more ergonomic to have some static data structure you're asserting the shape of and to do the assertion with a generic/recursive method like `cmp.Diff()` - which will error out early if the types don't match in the first place.

Reply:
If the AccessKey or SecretKey is not set, the AWS SDK loads the credentials with the default chain. In this chain there could (theoretically) be AWS credentials of the CI runner. Just saying, printing the actual AWS AccessKeyID and SecretAccessKey might not be the best idea.

In regards to `cmp.Diff()` : Do you mean something like recreating the s3Credentials struct from the S3 options and comparing that to the provided one?

Reply:
Yikes, wow. Is that the expected behavior for our code, even? Seems wild that their loader only exposes `LoadDefaultConfig()` and doesn't allow an explicit credential loader. 

Re: diff - yes, a field `expected` in the test struct that is either the `s3.Options` or the `s3.Credentials` (depending on what you found more useful to test) was my hope.

Reply:
Sorry, I understand how the default loading grabs the environment, I was just wondering if that was ever the expected deployment mechanism for Prow, since we provide an explicit API here. Omissions in config leading to implicit config loading might not be what everyone expects. Likely nothing we can do about it.